8282024: add EscapeAnalysis statistics under PrintOptoStatistics

aamarsh opened the pull request:
Escape Analysis and Scalar Replacement statistics were added when the -XX:+PrintOptoStatistics flag is set. All code is placed in `#ifndef Product` block, so this code is only run when creating a debug build. Using renaissance benchmark I ran a few tests to confirm that numbers were printing correctly. Below is an example run:

```
No escape = 263, Arg escape = 87, Global escape = 1628
Objects scalar replaced = 193, Monitor objects removed = 32, GC barriers removed = 38, Memory barriers removed = 225
```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8282024](https://bugs.openjdk.java.net/browse/JDK-8282024): add EscapeAnalysis statistics under PrintOptoStatistics


### Reviewers
 * [Xin Liu](https://openjdk.java.net/census#xliu) (@navyxliu - Committer)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8019/head:pull/8019` \
`$ git checkout pull/8019`

Update a local copy of the PR: \
`$ git checkout pull/8019` \
`$ git pull https://git.openjdk.java.net/jdk pull/8019/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8019`

View PR using the GUI difftool: \
`$ git pr show -t 8019`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8019.diff">https://git.openjdk.java.net/jdk/pull/8019.diff</a>

</details>

JohnTortugo commented on src/hotspot/share/opto/macro.cpp:
```diff
@@ -2334,6 +2341,11 @@ void PhaseMacroExpand::eliminate_macro_nodes() {
       if (n->is_AbstractLock()) {
         success = eliminate_locking_node(n->as_AbstractLock());
+        if(success) {
+          #ifndef PRODUCT
```
You can move the if inside the "ifndef"

JohnTortugo commented on src/hotspot/share/opto/macro.cpp:
```diff
@@ -2569,3 +2590,35 @@ bool PhaseMacroExpand::expand_macro_nodes() {
+  tty->print_cr("Number of memory barriers removed: %d", _memory_barriers_removed_counter);
+}
+
+int PhaseMacroExpand::countMemBar() {
```
I think the name of this method doesn't follow the coding style.

JohnTortugo commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -3739,6 +3772,34 @@ void ConnectionGraph::dump(GrowableArray<PointsToNode*>& ptnodes_worklist) {
+}
+
+void ConnectionGraph::escape_state_statistics(GrowableArray<JavaObjectNode*>& java_objects_worklist) {
+  for(int next = 0; next < java_objects_worklist.length(); ++next) {
```
NIT: missing space after `for`.

JohnTortugo commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -3739,6 +3772,34 @@ void ConnectionGraph::dump(GrowableArray<PointsToNode*>& ptnodes_worklist) {
+      if(ptn->escape_state() == PointsToNode::NoEscape) {
+        ConnectionGraph::_no_escape_counter++;
+      }
+      else if(ptn->escape_state() == PointsToNode::ArgEscape) {
```
NIT: missing space after the "ifs".

navyxliu commented:
hi, @aamarsh , 
Generally speaking,  the scalar replaced objects should be the subset of NonEscape objects. 

In your case,  escape/Objects scalar replaced is 124/172.  One potential explanation is that you encounter many boxing objects. Is this case? 

```
  // Eliminate boxing allocations which are not used
  // regardless scalar replacable status.
  bool boxing_alloc = C->eliminate_boxing() &&
                      tklass->klass()->is_instance_klass()  &&
                      tklass->klass()->as_instance_klass()->is_box_klass();
  if (!alloc->_is_scalar_replaceable && (!boxing_alloc || (res != NULL))) {
    return false;
  }
```


I see that you add  static counters to ConnectionGraph and PhaseMacroExpand. Do we need to use Atomic::add in case many C2 compiler threads run in parallel.   if you combine CICompilerCount and -Xcomp, can you see those counters are stable values? 

Even we won't completely mess up counters, it may lose values in multi-threaded environment.
```
PhaseMacroExpand::_objs_scalar_replaced_counter++;
```

aamarsh commented:
hi, @navyxliu 

Yes, you are correct that scalar replaced objects should be a subset of NonEscape objects. In my implementation I was resetting the NonEscape counter if escape analysis was called multiple times on a ConnectionGraph. Unfortunately I was forgetting to count the objects that had been scalar replaced in between escape analysis calls. I have updated my logic to support this and I updated the provided example. Thank you for catching that!

As for using static counters, the thought of a multi-threaded environment did come up. I looked at other static counters printed in `Compile::print_statistics()` for examples. None of the static counters that I found took into account a multi-threaded environment. Some of these counters include `SharedRuntime::_implicit_null_throws`, `PhaseCCP::_total_invokes`,  `PhaseCCP::_total_constants`, `PhaseOutput::total_nop_size`, and `PhasePeephole::total_peepholes`. But adding this logic most definitley does not hurt, so I did use the `Atomic` functions to load/store/add my static counters.

aamarsh commented:
hi, @navyxliu 

Yes, you are correct that scalar replaced objects should be a subset of NonEscape objects. In my implementation I was resetting the NonEscape counter if escape analysis was called multiple times on a ConnectionGraph. Unfortunately I was forgetting to count the objects that had been scalar replaced in between escape analysis calls. I have updated my logic to support this and I updated the provided example. Thank you for catching that!

As for using static counters, the thought of a multi-threaded environment did come up. I looked at other static counters printed in `Compile::print_statistics()` for examples. None of the static counters that I found took into account a multi-threaded environment. Some of these counters include `SharedRuntime::_implicit_null_throws`, `PhaseCCP::_total_invokes`,  `PhaseCCP::_total_constants`, `PhaseOutput::total_nop_size`, and `PhasePeephole::total_peepholes`. But adding this logic most definitley does not hurt, so I did use the `Atomic` functions to load/store/add my static counters.

navyxliu commented:
hi, @aamarsh ,
I understand that. I guess it wasn't easy to have portal atomic operations. The situation has improved since atomic.hpp is introduced. we can use it as easy as standard c++.  All those counters are in debug build, so we won't pay anything in release build. let's leave old counters for future work.

navyxliu commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -2188,11 +2190,17 @@ void Compile::Optimize() {
         igvn.set_delay_transform(false);
 
+#ifndef PRODUCT
+        congraph()->update_escape_state(Atomic::load(&PhaseMacroExpand::_objs_scalar_replaced_counter) - _prev_scalar_replaced);
```
I understand that you would like to add scalarized objects back to the snapshot, but this _objs_scalar_replaced_counter is a static counter as well. Two consecutive atomic loads are not helpful here because other C2 compiler threads can update it.

I think we can a member data ConnectionGraph::_prev_scalar_replaced and increment it in mexp.eliminate_macro_nodes();  we keep _objs_scalar_replaced_counter  but only atomic accumulate it at the end of ME phrase.

aamarsh replied:
@navyxliu I followed the general idea of your advice and created _local_scalar_replaced, which gets added in compiler.cpp.

navyxliu commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -95,15 +101,31 @@ void ConnectionGraph::do_analysis(Compile *C, PhaseIterGVN *igvn) {
   if (C->congraph() != NULL) {
     invocation = C->congraph()->_invocation + 1;
+#ifndef PRODUCT
+    // Reset counters when do_analysis is called again so objects are not double counted
```
This does not look right either.  3 atomic words can't give you a safe transaction. 
I understand that you want to dedup across Iterative EAs. Snapshot is certainly a general solution but more complex. 

In our case,  I feel all we need is to compensate those java objects which have been scalarized in previous iterations. One member data _prev_scalar_replaced, which remembers the number of scalarized objects, seems good enough.   you can carry over this variable from the old to the new ConnectionGraph and add it back when the EA iterations end.

aamarsh replied:
@navyxliu Again, to avoid this I added local counters to the compiler that reset with every do_analysis call and once EA is completed and the while loop is executed they are atomically added to the static counters.

navyxliu commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -3739,6 +3791,38 @@ void ConnectionGraph::dump(GrowableArray<PointsToNode*>& ptnodes_worklist) {
 }
 
+void ConnectionGraph::print_statistics() {
+  tty->print_cr("No escape: %d", Atomic::load(&_no_escape_counter));
```
This is just my suggestion. I would say that we keep each print_statistics() method oneliner.  
It will ease parsers.  in Compile::print_statistics, most of them follow the pattern: phase: counter1, couter2, ...
```  
  tty->print_cr("Peephole: peephole rules applied: %d",  _total_peepholes);
```

aamarsh replied:
@navyxliu Fixed! And I included a new example of how it prints above!

navyxliu commented on src/hotspot/share/opto/macro.cpp:
```diff
@@ -2569,3 +2591,35 @@ bool PhaseMacroExpand::expand_macro_nodes() {
+  tty->print_cr("Memory barriers removed = %d", Atomic::load(&_memory_barriers_removed_counter));
+}
+
+int PhaseMacroExpand::count_MemBar() {
```
I am not sure about this procedural. Even though you use Unique_Node_List, is it still possible to count the same membar multiple times?  or a backedge causes a infinitely loop? 

I think you can use Compile::identify_useful_nodes to collect all useful nodes and then count MemBar() nodes.

aamarsh replied:
@navyxliu Since no nodes are ever removed from the list, I think this issue is avoided. If we find a loop back to a node we have already searched and attempt to push this node at line 2620, it will not be added to the list because it already exists.

navyxliu replied:
nits: 
This function seems not to be part of 'PhaseMacroExpand', at least it could be a static member function.

vnkozlov replied:
You should check `PrintOptoStatistic` flag.

navyxliu commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -2204,6 +2209,13 @@ void Compile::Optimize() {
+
+#ifndef PRODUCT
+    Atomic::add(&ConnectionGraph::_no_escape_counter, total_scalar_replaced);
+    Atomic::add(&ConnectionGraph::_no_escape_counter, _local_no_escape_ctr);
```
you can merge two atomic operations into one.  Atomic::add(&ConnectionGraph::_no_escape_counter(&ConnectionGraph::_no_escape_counter, _local_no_escape_ctr + total_scalar_replaced); 

This is smart. last revision is the final revision.

aamarsh replied:
@navyxliu Thank you so much for your feedback, it is helpful for me :).

As per this issue, since no nodes are ever removed from the list, I think this issue is avoided. If we find a loop back to a node we have already searched and attempt to push this node at line 2620, it will not be added to the list because it already exists.

navyxliu reviewed (commented):
hi, 

Thank you for working on this. In generally, I think it looks good.  only small suggests left.

navyxliu commented:
It seems that you use git reset and push it forcely. That's why you only have one commit displayed in this PR. 
Good news is that Skara can still generate revisions. 

I think it's easier for people to keep track your changes incrementally. It's totally okay you keep committing to your branch.  github will squash them when it merges a PR. please refrain from push by force.

navyxliu commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -3739,6 +3773,34 @@ void ConnectionGraph::dump(GrowableArray<PointsToNode*>& ptnodes_worklist) {
+      else if (ptn->escape_state() == PointsToNode::ArgEscape) {
+        _compile->_local_arg_escape_ctr++;
+      }
+      else if (ptn->escape_state() == PointsToNode::GlobalEscape) {
```
"else if" style is not consistent with others.

JohnTortugo replied:
NIT: Move "else if" to line 3793.

navyxliu reviewed (approved):
LTGM. I am not a reviewer.  We need other reviewers to approve it.

vnkozlov commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -233,6 +243,9 @@ bool ConnectionGraph::compute_escape() {
 
   if (non_escaped_allocs_worklist.length() == 0) {
     _collecting = false;
+#ifndef PRODUCT
+    escape_state_statistics(java_objects_worklist);
+#endif
```
You can use `NOT_PRODUCT()` macro for one line which you have a lot in these changes.

JohnTortugo replied:
NIT: Use NOT_PRODUCT

vnkozlov commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -3739,6 +3773,34 @@ void ConnectionGraph::dump(GrowableArray<PointsToNode*>& ptnodes_worklist) {
+  tty->print_cr(" (EA executed in %7.2f seconds)", Atomic::load(&_time_elapsed) * 0.001);
+}
+
+void ConnectionGraph::escape_state_statistics(GrowableArray<JavaObjectNode*>& java_objects_worklist) {
```
You should check `PrintOptoStatistic` flag to avoid useless work.

vnkozlov commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -3739,6 +3773,34 @@ void ConnectionGraph::dump(GrowableArray<PointsToNode*>& ptnodes_worklist) {
+}
+
+void ConnectionGraph::escape_state_statistics(GrowableArray<JavaObjectNode*>& java_objects_worklist) {
+  _compile->_local_no_escape_ctr = 0;
+  _compile->_local_arg_escape_ctr = 0;
+  _compile->_local_global_escape_ctr = 0;
```
I don't think you need these `Compile` counters - make them local to update corresponding `ConnectionGraph` counters after loop. As I said these counters are not accurate anyway. Unless you want to track for which allocation which counter was recorded. Which I think will be over-kill because the most EA done with one iteration.

vnkozlov commented on src/hotspot/share/opto/macro.hpp:
```diff
@@ -200,14 +200,24 @@ class PhaseMacroExpand : public Phase {
+    static int _monitor_objects_removed_counter;
+    static int _GC_barriers_removed_counter;
+    static int _memory_barriers_removed_counter;
+    int _local_scalar_replaced;
```
You don't need `_local_scalar_replaced` value as I said in an other comment.

vnkozlov commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -116,6 +120,12 @@ void ConnectionGraph::do_analysis(Compile *C, PhaseIterGVN *igvn) {
     igvn->hash_delete(noop_null);
   }
+
+#ifndef PRODUCT
+  et.stop();
+  // casting jlong to long since Atomic needs Integral type
+  Atomic::add(&ConnectionGraph::_time_elapsed, (long)et.milliseconds());
+#endif
```
You should check `PrintOptoStatistic` flag.

vnkozlov reviewed (changes requested):
I have few comments.

JohnTortugo commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -2204,6 +2209,12 @@ void Compile::Optimize() {
       // by removing some allocations and/or locks.
     } while (progress);
+
+#ifndef PRODUCT
+    Atomic::add(&ConnectionGraph::_no_escape_counter, _local_no_escape_ctr + total_scalar_replaced);
+    Atomic::add(&ConnectionGraph::_arg_escape_counter, _local_arg_escape_ctr);
+    Atomic::add(&ConnectionGraph::_global_escape_counter, _local_global_escape_ctr);
+#endif
```
Thanks for the review @vnkozlov .

> These should be done inside ConnectionGraph - don't expose EA counters to an other class. You can use a static method in ConnectionGraph to do that.

I also didn't like this. The iterative EA and not having the last `congraph` available after the loop complicated a bit the logic here. We'll work to improve this code.

> _no_escape_counter, _local_no_escape_ctr + total_scalar_replaced is wrong. You are doubling number because total_scalar_replaced is part of _local_no_escape_ctr

I see your point here. As @navyxliu mentioned, the "local_scalar_replaced" is used as an adjustment to keep record of the objects scalar replaced in **previous iterations**. However, this approach incorrectly handles numbers in the **last iteration**. In the last iteration _local_no_escape_ctr is always correct and we don't need to  add the objects SR in that iteration.

navyxliu replied:
I see. this patch counts the scalarized objects twice in last iteration. need to subtract `mexp._local_scalar_replaced` .

aamarsh replied:
@vnkozlov Thank you for the feedback! I removed the complicated iterative feedback fix and simply left a comment in the output. My example output above is updated. I added the PrintOptoStatistics flag where it felt necessary. I also pushed my commits incrementally as opposed to squashing them and force pushing as @navyxliu suggested.

JohnTortugo commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -2204,6 +2207,7 @@ void Compile::Optimize() {
       // Try again if candidates exist and made progress
       // by removing some allocations and/or locks.
     } while (progress);
+
```
NIT: empty line.

JohnTortugo commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -2174,6 +2176,7 @@ void Compile::Optimize() {
       if (failing())  return;
     }
     bool progress;
+
```
NIT: empty line.

JohnTortugo reviewed (commented):
Some NITs.

vnkozlov commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -3739,6 +3767,32 @@ void ConnectionGraph::dump(GrowableArray<PointsToNode*>& ptnodes_worklist) {
+void ConnectionGraph::print_statistics() {
+  tty->print("No escape = %d, Arg escape = %d, Global escape = %d", Atomic::load(&_no_escape_counter), Atomic::load(&_arg_escape_counter), Atomic::load(&_global_escape_counter));
+  tty->print(" (EA executed in %7.2f seconds)", Atomic::load(&_time_elapsed) * 0.001);
+  tty->print_cr("   ** EA stats might be slightly off since objects might be double counted due to iterative EA **");
```
I don't think we need to print this warning. I suggest to put it as comment in this method.

vnkozlov reviewed (commented):
Thank you for addressing my comment. I have only one left.
I started testing.

vnkozlov reviewed (commented):
About time statistic. There is flag -XX:+CITime which prints times for all phases in C2 (and C1). For EA it is `_t_escapeAnalysis`. And I see difference in reported times (0.02 vs 0.059).

Your output:
```
(EA executed in    0.02 seconds)
```

CITime:
```
         Escape Analysis:       0.076 s
           Conn Graph:            0.059 s
           Macro Eliminate:       0.027 s
```

I simply run fastdebug VM with -Xcomp:
```
java -XX:+PrintOptoStatistics -XX:+CITime -Xcomp t
```

vnkozlov commented:
It is up to you but I would suggest to remove your change to collect time since we have it already with CITime.

navyxliu commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -3739,6 +3752,31 @@ void ConnectionGraph::dump(GrowableArray<PointsToNode*>& ptnodes_worklist) {
 }
 
+void ConnectionGraph::print_statistics() {
+  // EA stats might be slightly off since objects might be double counted due to iterative EA
```
I don't understand. your approach almost worked in last revision. you all need to do is to adjust for the last iteration. 

This revision drop it. I don't think it would be "slightly" off, even double counted is optimistic.  A java object will be counted repeat if iterEA iterates N times.   my concern is the final statistical counters become incomparable. 

Why did you drop snapshot approach in https://github.com/openjdk/jdk/pull/8019/commits/0805514aec4c3d0bd5ec935c089e315e4b37c7fa?

vnkozlov replied:
@navyxliu, please, explain what do you mean "snapshot approach" and suggest how we should do it.

navyxliu replied:
In previous revision, @aamarsh used 3 data members of Compile.  This tuple is a snapshot. 

```
_local_no_escape_ctr
_local_arg_escape_ctr
_local_global_escape_ctr
```

ConnectionGraph::escape_state_statistics() initializes them all zeros and categorize Java objects of the connection graph. 

here is the framework. 
```
   do {
        EscapeAnalysis();
        MacroExpand.eliminate_macro_nodes();      
    } while (progress?);

#ifndef PRODUCT
    Atomic::add(&ConnectionGraph::_no_escape_counter, _local_no_escape_ctr + total_scalar_replaced);
    Atomic::add(&ConnectionGraph::_arg_escape_counter, _local_arg_escape_ctr);
    Atomic::add(&ConnectionGraph::_global_escape_counter, _local_global_escape_ctr);
#endif
```
Both you and @JohnTortugo  pointed out that there was a bug in previous revision. We overlook that non-escaped objects are double-counted in last iteration. I think it's amendable.  

I call this snapshot approach because it uses the snapshot of last iteration to update global statistical counters.  all intermediate snapshots are drop. Allow me to write down @aamarsh 's approach. 

The number of Java object `JO` is from user-program by nature.  EscapeAnalysis breaks them down into 3 categories. non-escaped, arg-escaped and global escaped.  Without iterative EA, we can report this snapshot to statistical counters.

With iterative EA, a problem arise. Some objects elided in  previous iteration of MacroExpansion. if we use last snapshot, we have to add those eliminated java objects. Let's say iterative EA iterates N times in total. `E` is the number of eliminated java object from 1 to N-1 iterations (exclude the last iteration here).  

Since all eliminated objects must be non-escaped, so we add E back to _no_escape_counter. We account for all java objects of this compilation unit.  
```
    Atomic::add(&ConnectionGraph::_no_escape_counter, _local_no_escape_ctr + E);
    Atomic::add(&ConnectionGraph::_arg_escape_counter, _local_arg_escape_ctr);
    Atomic::add(&ConnectionGraph::_global_escape_counter, _local_global_escape_ctr);
```

In this way, We keep track of all java objects of this CU  for iterative EA. I think it's accurate. 
```
JO = _local_no_escape_ctr + E + _local_arg_escape_ctr  + _local_global_escape_ctr
```

I suggest to hoist the variable `PhaseMacroExpand mexp` out of loop and make it stateful to track E.

vnkozlov replied:
Got it. Thank for explaining - you simply used data from last iteration.
I think you can do similar with much less complexity if you used data from **first** iteration and current code change:
```
void ConnectionGraph::escape_state_statistics(GrowableArray<JavaObjectNode*>& java_objects_worklist) {
  if (!PrintOptoStatistics || (_invocation > 0)) { // Collect data only for first invocation
    return;
  }
  for (int next = 0; next < java_objects_worklist.length(); ++next) {
```

vnkozlov replied:
You need to revert `escape_state_statistics` back to non static for this of cause.

navyxliu replied:
so you use the snapshot of first iteration. We assume Iterative EA can elide some non-escaped objects but can't change any object's category.  I think it is correct and indeed much simpler!

JohnTortugo replied:
That was a great idea. Thank you all. I'll work with @aamarsh to do the changes.

navyxliu reviewed (approved):
LGTM. I am not a reviewer. we need other reviewers to approve this.

vnkozlov reviewed (approved):
Looks good. I verified output.

aamarsh commented:
@vnkozlov It looks like there are two Windows x64 tests failing and after a re-run they have been stuck on the task for 20 hours. The failure happened at:

tools/javac/Paths/MineField.sh 
tools/javac/Paths/wcMineField.sh 

with the message:

`STDERR:
Main.java:1: error: cannot find symbol
public class Main {public static void main(String[] a) {Lib.f();}}`

It does not appear that my code would effect these tests, but I don't want to integrate if the check are are not all passing.

vnkozlov commented:
@aamarsh  Do not worry. I see other latest PRs has the same failure. Your changes should not affect product version of JDK.

vnkozlov commented:
Give me time to quick test to make sure it builds on all our systems. I will sponsor after that.

JohnTortugo commented:
No worries. Thank you Vladimir!

vnkozlov commented:
/sponsor
